timemachine/fe/free_energy.py

```diff
@@ -267,6 +274,38 @@ def frames(self) -> List[StoredArrays]:
     def boxes(self) -> List[NDArray]:
         return [np.array(traj.boxes) for traj in self.trajectories]
 
+    def compute_u_kn(self) -> Tuple[NDArray, NDArray]:
```

Comment:
On the fence : might want to merge this function with https://github.com/proteneer/timemachine/blob/60d823020dfb0c7f0bbe4b25a4b9b8a62c9115cd/timemachine/fe/free_energy.py#L1065-L1073 , but may need to generalize signature. (in other function, we have a length-K container of permuted (x,box) states in `HREX[CoordsVelBox]`, here we have a length-K container of (xs, boxes) trajectories, ...)

Reply:
hmm, it seems like the common element is the batch evaluation of potentials over (x, box) pairs? Do you imagine `compute_u_kn` could make use of the sparse evaluation strategy that's implemented by `compute_potential_matrix`? Otherwise, it doesn't seem like much is shared (and extra logic that isn't shared, e.g. the translation between replica and state indices in `compute_potential_matrix`)